cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/TaskWorkerServiceLauncher.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.app.worker;
+
+import com.google.common.util.concurrent.AbstractIdleService;
+import com.google.inject.Inject;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.conf.SConfiguration;
+import io.cdap.cdap.common.utils.DirUtils;
+import io.cdap.cdap.master.spi.twill.StatefulDisk;
+import io.cdap.cdap.master.spi.twill.StatefulTwillPreparer;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.twill.api.ResourceSpecification;
+import org.apache.twill.api.TwillController;
+import org.apache.twill.api.TwillPreparer;
+import org.apache.twill.api.TwillRunner;
+import org.apache.twill.common.Threads;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.Writer;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Launches a pool of task workers.
+ */
+public class TaskWorkerServiceLauncher extends AbstractIdleService {
+  private static final Logger LOG = LoggerFactory.getLogger(TaskWorkerServiceLauncher.class);
+
+  private final CConfiguration cConf;
+  private final Configuration hConf;
+  private final SConfiguration sConf;
+
+  private final TwillRunner twillRunner;
+  private TwillController twillController;
+
+  @Inject
+  public TaskWorkerServiceLauncher(CConfiguration cConf, SConfiguration sConf, Configuration hConf,
+                                   TwillRunner twillRunner) {
+    this.cConf = cConf;
+    this.hConf = hConf;
+    this.sConf = sConf;
+    this.twillRunner = twillRunner;
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    LOG.info("Starting TaskWorkerServiceLauncher.");
+
+    launchTaskWorkerPool();
+
+    LOG.info("Starting TaskWorkerServiceLauncher has completed.");
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    LOG.info("Shutting down TaskWorkerServiceLauncher has completed.");
+  }
+
+  private void launchTaskWorkerPool() {
```

Comment:
We should periodically execute this logic to make sure the pool is up and running, right? See `DistributedPreviewManager`.

Reply:
Done!